### PR TITLE
YOLOv8-Ghost-P2 config

### PR DESCRIPTION
**Pull request by Burhan-Q**
Adds YOLOv8-Ghost-P2 config by @ExtReMLapin commented on #6749 

# Model testing results

> Pending

### Summary
🆕👻🕵️

Add a new YAML file `ultralytics/cfg/models/v8/yolov8-ghost-p2.yaml` that defines a lightweight YOLOv8 variant with GhostConv and C3Ghost modules. The file is part of a pull request that introduces new models and features to the ultralytics repository.

> _`YOLOv8-ghost-p2`_
> _A lighter detection model_
> _Autumn leaves falling_

### Walkthrough
* Add new YAML files to define YOLOv8 variants with GhostConv and C3Ghost modules ([link](https://github.com/ultralytics/ultralytics/pull/6797/files?diff=unified&w=0#diff-c02325469905c7e980bf8f9fec8fd4afd12d0de36bb989b2982f0d2ccfdcbd7cL1-R53),   )

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Introducing the new YOLOv8 Ghost-P2 model configuration 🚀

### 📊 Key Changes
- 🔧 Added a new `yolov8-ghost-p2.yaml` file with configuration details for different sizes of the YOLOv8 model variants (n, s, m, l, x).
- 🏗️ The configuration specifies the architecture for the YOLOv8-ghost backbone and YOLOv8-ghost-p2 head, detailing layers, convolutions, and concatenations.
- 💡 Features are enhanced with Ghost Convolution layers and C3Ghost modules for efficiency.
- 🚀 Defines the structure for P2 to P5 outputs in the model, indicating how multiple layers and scales are integrated.

### 🎯 Purpose & Impact
- 🎛️ Offers a scalable range of model sizes to balance between speed and accuracy depending on the application's requirements.
- 🚀 Improves inference speed and reduces model size by utilizing Ghost convolutions without sacrificing performance, beneficial for devices with limited computational resources.
- 🤖 Facilitates object detection with models at various complexity levels (from the smallest 'n' to the largest 'x'), serving a broad user base from hobbyists to professionals.
- 💥 Potential performance gains in object detection tasks across diverse deployment scenarios, including edge devices and cloud platforms.

### CODE REVIEW

**Comment by WangQvQ** [comment]
please see https://github.com/ultralytics/ultralytics/pull/6815

**Comment by glenn-jocher** [comment]
@Burhan-Q this looks good to merge, what do you think?

**Comment by Burhan-Q** [comment]
@glenn-jocher I think it's good to merge and then I'll work on figuring out how to cleanup the configs so we don't have so many YAML files. Don't have a full comparison yet (takes a while longer to train!) but did complete a single training run for YOLOv8s-P2-ghost.

```py
from ultralytics import YOLO

model = YOLO('yolov8s-p2-ghost.yaml')
_ = model.to('cuda')
_ = model.reset_weights()

results = model.train(data='VisDrone.yaml', epochs=100, batch=4)
```

<details>

<summary>Training final validation results</summary>

```
Ultralytics YOLOv8.0.215 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
YOLOv8s-p2-ghost summary (fused): 380 layers, 5307464 parameters, 0 gradients, 22.1 GFLOPs

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 
                   all        548      38759      0.453      0.359       0.36      0.211
            pedestrian        548       8844      0.476      0.415      0.421      0.192
                people        548       5125      0.471      0.306      0.331      0.129
               bicycle        548       1287      0.244       0.14      0.106     0.0451
                   car        548      14064      0.656      0.793      0.801      0.559
                   van        548       1975      0.473      0.426      0.425      0.292
                 truck        548        750      0.407      0.277      0.267      0.171
              tricycle        548       1045      0.448      0.205      0.225      0.124
       awning-tricycle        548        532      0.272      0.147      0.139     0.0895
                   bus        548        251      0.566      0.454      0.458      0.314
                 motor        548       4886      0.513      0.428      0.427      0.192
Speed: 0.2ms preprocess, 7.1ms inference, 0.0ms loss, 4.3ms postprocess per image
```

</details>

**Comment by glenn-jocher** [comment]
@Burhan-Q yeah we may need to rethink model building in future releases to modularize things better.

The current YAML configs are a huge leap vs the YOLOv5 structure, for YOLOv8 I created a single YAML that covers all scales and then smart filename handling to convert non-existent files like yolov8n.yaml into a command to load the base YAML and assign the 'n' scale to it.

With that change I reduced the outstanding model YAMLs by 5x, which was a great breakthrough. But there are arguments to be had for more modular methods like i.e. swapping in different backbones like ResNet101 onto a YOLOv8 detect head etc. It's complicated but would be nice to offer something like that in the future.

**Comment by Burhan-Q** [comment]
Purely for the sake of completeness, the results for YOLOv8s-P2 from scratch.

```py
from ultralytics import YOLO

model = YOLO('yolov8s-p2.yaml')
_ = model.to('cuda')
_ = model.reset_weights()

results = model.train(data='VisDrone.yaml', epochs=100, batch=4)
```

<details>

<summary>Training final validation results</summary>

```
Ultralytics YOLOv8.0.215 🚀 Python-3.10.12 torch-2.1.0+cu121 CUDA:0 (NVIDIA GeForce RTX 2060, 5924MiB)
YOLOv8s-p2 summary (fused): 207 layers, 10629048 parameters, 0 gradients, 36.7 GFLOPs

                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 
                   all        548      38759      0.489      0.395      0.397      0.238
            pedestrian        548       8844      0.523      0.445      0.461      0.218
                people        548       5125      0.535       0.33      0.373      0.151
               bicycle        548       1287      0.261       0.18      0.124     0.0547
                   car        548      14064      0.693      0.803      0.816      0.582
                   van        548       1975       0.52      0.467      0.467      0.328
                 truck        548        750      0.468      0.341      0.327      0.218
              tricycle        548       1045      0.466      0.247      0.261      0.149
       awning-tricycle        548        532      0.304      0.195      0.167      0.109
                   bus        548        251      0.584      0.463      0.497      0.356
                 motor        548       4886      0.536      0.477      0.475      0.219
Speed: 0.1ms preprocess, 4.5ms inference, 0.0ms loss, 7.0ms postprocess per image
```

</details>

|    Config                        | Weights file | Size (MB) |
|:--------------------------|:-------------:|:-----------:|
| yolov8s-p2.yaml            | `best.pt`       | 10.58       |
| yolov8s-p2-ghost.yaml | `best.pt`       | 20.64       |
